File: java/src/test/java/money_problem/domain/BankProperties.java
package money_problem.domain;

import com.pholser.junit.quickcheck.Property;
import com.pholser.junit.quickcheck.runner.JUnitQuickcheck;
import io.vavr.Tuple;
import io.vavr.Tuple2;
import io.vavr.collection.Map;
import io.vavr.control.Either;
import org.assertj.core.api.AbstractDoubleAssert;
import org.assertj.core.api.Assertions;
import org.assertj.core.data.Offset;
import org.junit.Test;
import org.junit.runner.RunWith;

import static io.vavr.collection.HashMap.of;
import static money_problem.domain.Currency.*;
import static org.assertj.vavr.api.VavrAssertions.assertThat;

@RunWith(JUnitQuickcheck.class)
public class BankProperties {
    private final Bank bank;
    private final Map<Tuple2<Currency, Currency>, Double> exchangeRates =
            of(
                    Tuple.of(EUR, USD), 1.2,
                    Tuple.of(USD, EUR), 0.82,
                    Tuple.of(USD, KRW), 1100d,
                    Tuple.of(KRW, USD), 0.0009,
                    Tuple.of(EUR, KRW), 1344d,
                    Tuple.of(KRW, EUR), 0.00073
            );

    public BankProperties() {
        this.bank = createBank();
    }

    @Property
    public void convertInSameCurrencyShouldReturnOriginalMoney(double originalAmount, Currency currency) {
        var originalMoney = new Money(originalAmount, currency);

        assertThat(bank.convert(originalMoney, currency))
                .containsOnRight(originalMoney);
    }

    @Property
    public void roundTrippingInDifferentCurrencies(double originalAmount, Currency from, Currency to) {
        assertThat(roundTripConvert(originalAmount, from, to))
                .hasRightValueSatisfying(money ->
                        assertThatAmountAreClosed(originalAmount, money)
                );
    }

    @Test
    public void roundTripInError() {
        var originalAmount = 9.051E-11;
        assertThat(roundTripConvert(originalAmount, EUR, USD))
                .hasRightValueSatisfying(money -> assertThatAmountAreClosed(originalAmount, money));
    }

    private AbstractDoubleAssert<?> assertThatAmountAreClosed(double originalAmount, Money money) {
        return Assertions.assertThat(money.amount())
                .isCloseTo(originalAmount, Offset.offset(0.1));
    }

    private Either<String, Money> roundTripConvert(double originalAmount, Currency from, Currency to) {
        return bank.convert(new Money(originalAmount, from), to)
                .flatMap(convertedMoney -> bank.convert(convertedMoney, from));
    }

    private Bank createBank() {
        return exchangeRates
                .foldLeft(newBank(), (bank, exchangeRate) -> bank.addExchangeRate(exchangeRate._1._1, exchangeRate._1._2, exchangeRate._2));
    }

    private Bank newBank() {
        var firstEntry = exchangeRates.head();
        return Bank.withExchangeRate(firstEntry._1()._1, firstEntry._1()._2, firstEntry._2);
    }
}

File: scala/src/test/scala/domain/BankProperties.scala
package domain

import domain.BankProperties.{bank, currencyGenerator, roundTripConvert}
import domain.Currency._
import org.scalacheck.Prop.forAll
import org.scalacheck.{Arbitrary, Gen}
import org.scalactic.Tolerance.convertNumericToPlusOrMinusWrapper
import org.scalatest.EitherValues
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.scalacheck.Checkers

class BankProperties extends AnyFunSuite with Checkers with EitherValues {
  test("convert in same currency should return original Money") {
    forAll { (originalAmount: Double, currency: Currency) =>
      {
        val originalMoney = Money(originalAmount, currency)
        bank.convert(originalMoney, currency).value == originalMoney
      }
    }.check()
  }

  test("round tripping in different currencies") {
    forAll { (originalAmount: Double, from: Currency, to: Currency) =>
      {
        roundTripConvert(originalAmount, from, to).value == Money(
          originalAmount,
          from
        )
      }
    }.check()
  }

  test("round trip in error") {
    val originalAmount = -2.1645893211081448e69
    val roundTripAmount = roundTripConvert(originalAmount, EUR, USD).value

    assert(roundTripAmount.amount === originalAmount +- 0.01)
  }

  private def toleranceFor(originalAmount: Double): Double =
    Math.abs(0.01 * originalAmount)
}

object BankProperties {
  private val exchangeRates: Map[(Currency, Currency), Double] = Map(
    (EUR, USD) -> 1.2,
    (USD, EUR) -> 0.82,
    (USD, KRW) -> 1100d,
    (KRW, USD) -> 0.0009,
    (EUR, KRW) -> 1344d,
    (KRW, EUR) -> 0.00073
  )

  private val bank: Bank = createBank()

  private def newBank(): Bank = {
    val firstEntry = exchangeRates.head
    Bank.withExchangeRate(firstEntry._1._1, firstEntry._1._2, firstEntry._2)
  }

  private def createBank(): Bank =
    exchangeRates
      .drop(1)
      .foldLeft(newBank()) {
        case (bank, ((from, to), rate)) => bank.addExchangeRate(from, to, rate)
      }

  private def roundTripConvert(
      originalAmount: Double,
      from: Currency,
      to: Currency
  ): Either[String, Money] = {
    bank
      .convert(Money(originalAmount, from), to)
      .flatMap(convertedMoney => bank.convert(convertedMoney, from))
  }

  implicit def currencyGenerator: Arbitrary[Currency] =
    Arbitrary {
      Gen.oneOf(Currency.values.toSeq)
    }
}
